application/core/LSYii_Application.php

```diff
@@ -514,6 +508,78 @@ public function getTwigCustomExtensionsConfig($sUsertwigextensionrootdir, $aAppl
         return $aApplicationConfig;
     }
 
+    /**
+     * Handles specific exception cases, like "user friendly" exceptions and exceptions on ajax requests.
+     *
+     * @param CException $exception
+     * @return void
+     */
+    private function handleSpecificExceptions($exception)
+    {
+        if (
+            Yii::app()->request->isAjaxRequest &&
+            $exception instanceof CHttpException
+        ) {
+            $this->outputJsonError($exception);
+        } elseif ($exception instanceof LSUserException) {
+            $this->handleFriendlyException($exception);
+        }
+    }
+
+    /**
+     * Handles "friendly" exceptions by setting a flash message and redirecting.
+     * If the exception doesn't specify a redirect URL, the referrer is used.
+     *
+     * @param array $error
+     * @param LSUserException $exception
+     * @return void
+     */
+    private function handleFriendlyException($exception)
+    {
+        $message = "<p>" . $exception->getMessage() . "</p>";
+        $message .= $exception->getDetailedErrorSummary();
+        Yii::app()->setFlashMessage($message, 'error');
+        if ($exception->getRedirectUrl() != null) {
+            $redirectTo = $exception->getRedirectUrl();
+        } else {
+            $redirectTo = Yii::app()->request->urlReferrer;
+        }
+        Yii::app()->request->redirect($redirectTo);
+    }
+
+    /**
+     * Outputs an exception as JSON.
+     *
+     * @param CHttpException $exception
+     * @return void
+     */
+    private function outputJsonError($exception)
+    {
+        $outputData = [
+            'success' => false,
+            'message' => $exception->getMessage(),
+        ];
+        if ($exception instanceof LSUserException) {
+            if ($exception->getRedirectUrl() != null) {
+                $outputData['redirectTo'] = $exception->getRedirectUrl();
+            }
+            if ($exception->getNoReload() != null) {
+                $outputData['noReload'] = $exception->getNoReload();
+            }
+            $errors = $exception->getDetailedErrors();
+
+            // Add the detailed errors to the message, so simple handlers can just show it.
+            $outputData['message'] .= "<br>" . implode("<br>", $errors);
```

Comment:
Possible to use same HTML than https://github.com/LimeSurvey/LimeSurvey/pull/2439/files#diff-830594dcfebb3cf724c6c04a8b9791bf841a954d95a84614481007be4568a060R100 ?

Reply:
Oh … thing need JS improvment here …